### ISSUE TITLE
[C#] Implement new Concept Exercise: memory-allocation

### ISSUE THREAD

**Issue by ErikSchierboom** [exercism/csharp#1444]
This issue describes how to implement the `memory-allocation` concept exercise for the C# track.

## Getting started

**Please please please read the docs before starting.** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. So, before diving into the implementation, please read up on the following documents:

- [The features of v3](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md).
- [Rationale for v3](https://github.com/exercism/v3/blob/master/docs/rationale-for-v3.md).
- [What are concept exercise and how they are structured?](https://github.com/exercism/v3/blob/master/docs/features-of-v3.md)

Please also watch the following video:

- [The Anatomy of a Concept Exercise](https://www.youtube.com/watch?v=gkbBqd7hPrA).

## Goal

The goal of this exercise is to teach the student the Concept of Memory Allocation in C#.

## Learning objectives

- Know what the stack and heap are.
- Know what the garbage collector is.
- Know how the stack and heap are different.
- Know what the large object heap is.
- Know what reference and value types are.
- Know how reference and value types relate to the stack and heap.

## Out of scope

- Memory locality.
- Memory pooling.
- `Span<T>` and `Memory<T>`
- `stackalloc`
- `unsafe` code.
- `fixed` statements.
- Defensive copies and `readonly` semantics.
- Discussion of profiling tools (referring to them is fine).

## Concepts

This Concepts Exercise's Concepts are:

- `memory-allocation`: know what the stack and heap are; know what the garbage collector is; know how the stack and heap are different; know what the large object heap is; know what reference and value types are; know how reference and value types relate to the stack and heap.

Any data types used in this exercise (e.g. `strings`) should also be added as prerequisites.

## Prequisites

This Concept Exercise's prerequisites Concepts are:

- `classes`: know how to define and work with classes.
- `structs`: know how to define and work with structs.

### Hints

- [Value types][value-types]: introduction to value types.
- [Reference types][reference-types]: introduction to reference types.
- [Stack versus heap][stack-vs-heap]: how the stack and heap work.

### After

- [Value types][value-types]: introduction to value types.
- [Reference types][reference-types]: introduction to reference types.
- [Stack versus heap][stack-vs-heap]: how the stack and heap work.
- [Value types vs reference types][value-types-vs-reference-types]: in-depth comparison between value and reference types
- [Garbage collection][garbage-collection]: how garbage collection works in C#.
- [Large Object Heap][large-object-heap]: what the Large Object Heap is.
- [Garbage collection performance][garbage-collection-performance]: in-depth look into garbage collection performance.

## Representer

This exercise does not require any specific representation logic to be added to the [representer][representer].

## Analyzer

This exercise does not require any specific analyzer logic to be added to the [analyzer][analyzer].

## Implementing

To implement this exercise, please [follow these instructions](https://github.com/exercism/v3/blob/master/languages/csharp/reference/implementing-a-concept-exercise.md).

## Help

If you have any questions while implementing the exercise, please post the questions as comments in this issue.

[how-to-implement-a-concept-exercise]: https://github.com/exercism/v3/blob/master/docs/maintainers/generic-how-to-implement-a-concept-exercise.md
[implemented-exercises]: https://github.com/exercism/v3/tree/master/languages/csharp/exercises/concept/README.md#implemented-exercises
[reference]: https://github.com/exercism/v3/blob/master/languages/csharp/reference/README.md#reference-docs
[reference-doc]: https://github.com/exercism/v3/blob/master/reference/concepts/performance.md
[reference-example]: https://github.com/exercism/v3/blob/master/reference/types/string.md#implementations
[analyzer]: https://github.com/exercism/csharp-analyzer
[representer]: https://github.com/exercism/csharp-representer
[exercise-example]: https://github.com/exercism/v3/tree/master/languages/csharp/exercises/concept/numbers-floating-point
[design-example]: https://github.com/exercism/v3/blob/master/languages/csharp/exercises/concept/numbers/.meta/design.md
[config.json-example]: https://github.com/exercism/v3/blob/master/languages/csharp/exercises/concept/numbers/.meta/config.json
[concept-exercises]: https://github.com/exercism/v3/blob/master/docs/concept-exercises.md
[value-types]: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/value-types
[reference-types]: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/reference-types
[stack-vs-heap]: https://www.c-sharpcorner.com/article/C-Sharp-heaping-vs-stacking-in-net-part-i/
[garbage-collection]: https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/fundamentals
[value-types-vs-reference-types]: https://adamsitnik.com/Value-Types-vs-Reference-Types/
[large-object-heap]: https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/large-object-heap
[garbage-collection-performance]: https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/performance
